### DISCUSSION TITLE
Standalone php cli?

### DISCUSSION THREAD

**Discussion by crazywhalecc** [crazywhalecc/static-php-cli#58] [Ideas]
Making a binary that not static but only linking glibc. It can be used as a normal binary to add other dynamic linked extension, such as xdebug, source guardian, libffi, etc.

This is only for linux OS discussion. Is it good for forking repo to make a `standalone-php-cli`?

**Comment by TNT-ma** [comment]
如果有一份加密PHP代码的 第三方扩展源代码(c语言)  能不能编译到 二进制PHP文件内？ 我觉得二进制 +  加密 缺一不可

**Reply by crazywhalecc** [reply]
这种理论上可以，但是也得看它本身有没有依赖不能静态编译的扩展。而且不出所料这类在代码解析上动刀的扩展大概率是要patch的。

**Reply by TNT-ma** [reply]
https://gitee.com/lfveeker/tonyenc
你帮忙看看这个源代码是否有可能（在linux上编译到PHP二进制内）
几个月没有搞PHP 了之前一直想二进制 加密 没有搞定。最近一直学习GO ，但是还是觉得PHP搞web项目更快捷。

**Reply by crazywhalecc** [reply]
我加到待完成列表，后面有时间试一下。

**Reply by TNT-ma** [reply]
😃迫切需要 可付费 打赏

**Reply by TNT-ma** [reply]
老大 我想付费定制这个加密扩展，业务原因最近的项目急需要这个加密文件扩展，能否私信联系一下

**Reply by crazywhalecc** [reply]
我最近时间也比较少，最早周末时候可以看看能不能实现。不过不一定能适配，如果不能加入，我会回复你。

**Reply by TNT-ma** [reply]
如果不能适配，是否可以按静态编译所需要求，写一个用于加密源代码的扩展？没有找到相关的文章或案例。

**Reply by crazywhalecc** [reply]
测试了下可以适配 tonyenc，但是 tonyenc 源自 gitee，众所周知国内的 Gitee 下载不是很稳，可能需要一个稳定的 Fork 版放在 GitHub。

**Reply by TNT-ma** [reply]
https://github.com/TNT-ma/tonyenc
我Fork  了一个

**Reply by crazywhalecc** [reply]
目前可以编译到静态 PHP 二进制内，但是使用 micro 还是有问题，可能还需要魔改一下 tonyenc 或者 phpmicro 的源码。

**Reply by TNT-ma** [reply]
二进制和 micro 居然不通用吗

**Reply by crazywhalecc** [reply]
那肯定，cli 是直接把文件扔给 zend 引擎，micro 还要裁切掉二进制后面的代码，然后裸执行代码，总之就是运行流程不是很一样。

**Reply by TNT-ma** [reply]
坐等 micro  支持

**Reply by crazywhalecc** [reply]
😢 那估计得等 dixyes 大佬或者懂 PHP 底层的大佬研究一下了，否则我又得回到大学时候啃 PHP 底层设计那本书了，或者如果 tonyenc 的维护者了解一些的话，也可以，主要就是 `cgi_compile_file` 和 micro 的文件加载入口问题。

**Reply by TNT-ma** [reply]
好吧 我去找一下  tonyenc  作者的帮助

**Reply by TNT-ma** [reply]
真是遗憾tonyenc 作者并不懂micro是什么，PHP静态编译他也不懂。只会写个扩展 .

**Reply by TNT-ma** [reply]
有没有相关的技术文章可以学习一下，给我推荐一下，最近没事准备研究一下 搞定这个 tonyenc  到micro 去

**Reply by crazywhalecc** [reply]
我家里有一本《[PHP 7底层设计与源码实现](https://item.jd.com/12355605.html)》，但是是 PHP7 的，很多内容我简单看了下和 PHP8 已经有出入了。基础的数据结构什么的还是通的，可参考。我没找到 PHP8 相关的书，可能纯英文的有，中文版这个可能是唯一一本了。

**Reply by TNT-ma** [reply]
处处都是坎 这纯英文那没戏了

**Comment by deltapath-eric** [comment]
May I know update status for this discussion?
As I think that idea is great which to make the library able to use the .so file by specifying in php.ini

like ioncube extension

**Comment by crazywhalecc** [comment]
Fixed by #592
